go-runtime/schema/config/analyzer.go

```diff
@@ -0,0 +1,112 @@
+package config
+
+import (
+	"go/ast"
+	"go/types"
+	"strconv"
+
+	"github.com/TBD54566975/golang-tools/go/analysis"
+	"github.com/TBD54566975/golang-tools/go/analysis/passes/inspect"
+	"github.com/TBD54566975/golang-tools/go/ast/inspector"
+
+	"github.com/TBD54566975/ftl/go-runtime/schema/common"
+)
+
+// Extractor extracts config values relating to another decl, e.g. database configurations associated with a
+// database decl.
+//
+// Configs follow a pattern where they implement an interface, like `ftl.DatabaseConfig`.
+// We extract values by looking at known receiver methods. For example:
+//
+//	type FooConfig struct{}
+//
+//	func (f FooConfig) Name() string {
+//	    return "foo"
+//	}
+//
+// From this, we'd extract the "foo" value as the database name for `FooConfig`.
+var Extractor = common.NewExtractor("config", (*Fact)(nil), Extract)
+
+type Tag struct{} // Tag uniquely identifies the fact type for this extractor.
+type Fact = common.DefaultFact[Tag]
+
+func Extract(pass *analysis.Pass) (interface{}, error) {
+	in := pass.ResultOf[inspect.Analyzer].(*inspector.Inspector) //nolint:forcetypeassert
+	nodeFilter := []ast.Node{
+		(*ast.FuncDecl)(nil),
+	}
+
+	in.Preorder(nodeFilter, func(n ast.Node) {
+		fn := n.(*ast.FuncDecl) //nolint:forcetypeassert
+
+		// skip if there is no receiver
+		if fn.Recv == nil || len(fn.Recv.List) == 0 {
+			return
+		}
+
+		// handle both pointer and non-pointer receivers
+		var ident *ast.Ident
+		var ok bool
+		switch expr := fn.Recv.List[0].Type.(type) {
+		case *ast.StarExpr:
+			ident, ok = expr.X.(*ast.Ident)
+			if !ok {
+				return
+			}
+		case *ast.Ident:
+			ident = expr
+		default:
+			return
+		}
+
+		recType := pass.TypesInfo.TypeOf(ident)
+		if recType == nil {
+			return
+		}
+
+		// receiver implements ftl.DatabaseConfig
+		if common.IsDatabaseConfigType(pass, recType) {
+			extractDatabaseConfig(pass, getDBType(pass, ident, recType), fn, ident)
+		}
+
+	})
+	return common.NewExtractorResult(pass), nil
+}
+
+func extractDatabaseConfig(pass *analysis.Pass, dbType common.DatabaseType, fn *ast.FuncDecl, receiver *ast.Ident) {
+	obj, ok := common.GetObjectForNode(pass.TypesInfo, receiver).Get()
+	if !ok {
+		return
+	}
+	if len(fn.Body.List) == 0 {
+		return
+	}
+	returnStmt, ok := fn.Body.List[0].(*ast.ReturnStmt)
+	if !ok {
+		return
+	}
+
+	switch fn.Name.Name {
+	case "Name":
+		lit, ok := returnStmt.Results[0].(*ast.BasicLit)
+		if !ok {
+			return
+		}
+		name, err := strconv.Unquote(lit.Value)
+		if err != nil {
+			return
```

Comment:
It seems like all the returns in this function should also be surfacing errors to the user? Are they surfaced elsewhere?

Reply:
this will visit every receiver method and evaluate them separately, so if you had some arbitrary method (`func (MyDbConfig) Blah()`) we wouldn't want to error here, instead just skip it

Reply:
we could error explicitly in the `case "Name":` switch though, to your point. will add